### PR TITLE
react: new design for report header

### LINKED ISSUE #1592: react: adopt styling from cucumber reports
Current:

image

https://reports.cucumber.io/

image

Will need to:

- [x] Get #1302 updated and merged first
- [x] Copy styling from Cucumber reports for:
  - [x] Statuses summary (counts and colours)
  - [x] Execution summary (stats, timing, meta stuff)
- [x] Come up with some good styling for the ci and git stuff
- [x] Come up with some good styling for the search/filter bit

This is also a chance to finish modularising the styling and finally remove the global stylesheet - this should hopefully make it a bit easier for people to use these components in React apps.

### PR DESCRIPTION
**Pull request by davidjgoss**
## Summary

Update the report header components to have a new look and feel.

## Details

The styling for the statuses summary and execution summary is more or less copied from Cucumber Reports, but also includes styling for CI and Git stuff. The styling for the search bar is new - feedback welcome!

We talked a while ago about maybe bringing in Tailwind; I decided not to do that at least for now, partly because the styling of these things was not too arduous anyway, and partly because on reflection it seems like a lot to impose a CSS framework on someone trying to use a targeted component library. Again, feedback welcome.

This PR also does a bunch of other things (sorry):

- Finally removes the global stylesheet from `@cucumber/react`, so all styles are colocated with their components. This includes repurposing the `CustomRendering` component as a more generic `CucumberReact` component which is essentially a root component that provides baseline styles and where you can (optionally) activate themes and/or custom rendering
- Finishes making the "dark" theme so it works for the whole report, also adds a "auto" theme which is light or dark depending on OS preference - this is now used in the html formatter
- Adds a bunch of tests and stories for components that didn't have them
- Decouples the search stuff from the cucumber/gherkin/envelope queries stuff so there are targeted hooks and wrapper components for each
- Leans on `date-fns` for date/duration formatting as much as possible
- Fixes and tweaks a few general styling concerns (e.g. consistently use `em` for type-related units)
- Generally tries to slim down the logic in `FilteredResults` so more is taken care of by other functions/hooks etc

## Motivation and Context

Fixes #1592.

## How Has This Been Tested?

Expanded tests with react testing library, also tested HTML formatter output by hand.

## Screenshots (if appropriate):

Attached (in a zip) an output from html formatter (open it and switch your computer to dark mode to see the auto theme). [examples-tables.feature.html.zip](https://github.com/cucumber/common/files/7807884/examples-tables.feature.html.zip)

Storybook is deployed here for convenience https://confident-bell-07b8ce.netlify.app/?path=/story/app-filteredresults--default

image

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Checklist:

- [ ] The change has been ported to Java.
- [ ] The change has been ported to Ruby.
- [ ] The change has been ported to JavaScript.
- [ ] The change has been ported to Go.
- [ ] The change has been ported to .NET.
- [x] I've added tests for my code.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have updated the CHANGELOG accordingly.

### CODE REVIEW

**Review by aurelien-reeves — COMMENTED** [review]
Wow, this is really impressive! O.o

I look a little bit at the repo to see how the new components look like. It seems pretty good actually!

Regarding the search bar, would it be possible to complete the loop symbol by beginning the placeholder with a verb or a short phrase with a verb before the examples? Like 'Search' or 'Search the report'? Or something like this?

File: react/javascript/src/components/app/OSIcon.tsx [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import React, { VoidFunctionComponent } from 'react'
+import { Windows } from './icons/Windows'
+import { Linux } from './icons/Linux'
+import { MacOS } from './icons/MacOS'
+
+export const OSIcon: VoidFunctionComponent<{ name: string }> = ({ name }) => {
+  // TODO need to pattern match more permissively here
```
Comment (davidjgoss) [inline react/javascript/src/components/app/OSIcon.tsx:7]: I'm aware this isn't good enough. Are you able to share what's in Cucumber Reports for this?

File: react/javascript/src/components/app/RuntimeIcon.tsx [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import React, { VoidFunctionComponent } from 'react'
+import { NodeJs } from './icons/NodeJs'
+import { Jvm } from './icons/Jvm'
+import { Ruby } from './icons/Ruby'
+
+export const RuntimeIcon: VoidFunctionComponent<{ name: string }> = ({ name }) => {
+  // TODO need to pattern match more permissively here
```
Comment (davidjgoss) [inline react/javascript/src/components/app/RuntimeIcon.tsx:7]: As other comment, needs to be better, do we have prior art?

**Comment by davidjgoss** [comment]
@aurelien-reeves I've tweaked the wording, thanks for calling that out.

File: react/javascript/src/components/app/OSIcon.tsx (hunk 1 above)
Comment (aslakhellesoy) [reply react/javascript/src/components/app/OSIcon.tsx:7]: Pretty close. Here:

```tsx
  if (!name) {
    return <span>Unknown OS</span>
  }

  if (name.match(/windows|win32/i)) {
    return <Windows />
  }
  if (name.match(/(darwin|mac)/i)) {
    return <MacOS />
  }
  if (name.match(/linux/i)) {
    return <Linux />
  }

  return <span>{name}</span>
```

File: react/javascript/src/components/app/RuntimeIcon.tsx (hunk 2 above)
Comment (aslakhellesoy) [reply react/javascript/src/components/app/RuntimeIcon.tsx:7]:
```tsx
  if (!name) {
    return <span>Unknown Runtime</span>
  }

  if (name.match(/(oracle|openjdk|java)/i)) {
    return <Java />
  }
  if (name.match(/ruby/i)) {
    return <Ruby />
  }
  if (name.match(/node/i)) {
    return <JavaScript />
  }

  return <span>{name}</span>
```

File: react/javascript/src/components/app/ExecutionSummary.tsx [hunk 3]
```diff
@@ -1,86 +1,127 @@
...
-              </td>
-            </tr>
-          )}
-          {meta.implementation && <Product name="Implementation" product={meta.implementation} />}
-          {meta.runtime && <Product name="Runtime" product={meta.runtime} />}
-          {meta.os && <Product name="OS" product={meta.os} />}
-          {meta.cpu && <Product name="CPU" product={meta.cpu} />}
-        </tbody>
-      </table>
+    <div className={styles.backdrop}>
```
Comment (aslakhellesoy) [inline react/javascript/src/components/app/ExecutionSummary.tsx:62]: I really like that all the `classNames` are external to the tsx! I suppose that means consumers that use tailwind for the rest of the app can define those classes like this?

```css
.list {
  @apply ... tailwind classes ...
}
```

File: react/javascript/src/components/app/SearchBar.tsx [hunk 4]
```diff
@@ -1,85 +1,84 @@
...
-          <FontAwesomeIcon icon={faSearch} />
-        </button>
+        <small className={styles.searchHelp}>
+          You can search with plain text or{' '}
+          <a
+            href="https://cucumber.io/docs/cucumber/api/#tag-expressions"
+            target="_blank"
+            rel="noreferrer"
+          >
+            Cucumber Tag Expressions
```
Comment (aslakhellesoy) [inline react/javascript/src/components/app/SearchBar.tsx:51]: I tried `not @passing` and that highlighted `@passing`. Does it really support tag *expressions*, or just tags?

File: react/javascript/test/components/app/ExecutionSummaryTest.tsx [hunk 5]
```diff
@@ -0,0 +1,128 @@
+import React from 'react'
+import { render } from '@testing-library/react'
```
Comment (aslakhellesoy) [inline react/javascript/test/components/app/ExecutionSummaryTest.tsx:2]: Great that you brought this in!

**Review by aslakhellesoy — APPROVED** [review]
Impressive work @davidjgoss!! I've left a few comments (outside the review) about the case statements for icons. Fix that and good to merge.

File: react/javascript/src/components/app/SearchBar.tsx (hunk 4 above)
Comment (davidjgoss) [reply react/javascript/src/components/app/SearchBar.tsx:51]: Good question - I didn't touch any of that as part of this. I'll have a look and raise a separate issue.

File: react/javascript/src/components/app/ExecutionSummary.tsx (hunk 3 above)
Comment (davidjgoss) [reply react/javascript/src/components/app/ExecutionSummary.tsx:62]: For the builtin styling, the class name that actually comes out will be a unique generated string - this is a nice feature of CSS modules gets around the global namespace problem of CSS (esp when we're in someone else's app).

For components with custom rendering support, when users provide their own class names they can be whatever they want, including class names that reference a separate stylesheet, ones imported from another CSS module, or even the tailwind utility class names directly.

```tsx
<CucumberReact
        customRendering={{
          Tags: {
            tags: 'custom-list-class',
            tag: 'custom-item-class',
          },
        }}
      >
```

(I haven't actually made `<ExecutionSummary/>` et all customisable as part of this - happy to do another PR if you're going to need that though!)

**Pull request merged**